lib/class-wp-theme-json-resolver-gutenberg.php

```diff
@@ -286,74 +295,91 @@ public static function get_theme_data( $deprecated = array(), $options = array()
-			}
-			if ( current_theme_supports( 'experimental-link-color' ) ) {
-				_doing_it_wrong(
-					current_theme_supports( 'experimental-link-color' ),
-					__( '`experimental-link-color` is no longer supported. Use `link-color` instead.', 'gutenberg' ),
-					'6.3.0'
-				);
-			}
+				$default_gradients = false;
+				if ( current_theme_supports( 'default-gradient-presets' ) ) {
+					$default_gradients = true;
+				}
+				if ( ! isset( $theme_support_data['settings']['color']['gradients'] ) ) {
+					// If the theme does not have any gradients, we still want to show the core ones.
+					$default_gradients = true;
+				}
+				$theme_support_data['settings']['color']['defaultGradients'] = $default_gradients;
+
+				// Allow themes to enable all border settings via theme_support.
+				if ( current_theme_supports( 'border' ) ) {
+					$theme_support_data['settings']['border']['color']  = true;
+					$theme_support_data['settings']['border']['radius'] = true;
+					$theme_support_data['settings']['border']['style']  = true;
+					$theme_support_data['settings']['border']['width']  = true;
+				}
+
+				// Allow themes to enable link colors via theme_support.
+				if ( current_theme_supports( 'link-color' ) ) {
+					$theme_support_data['settings']['color']['link'] = true;
+				}
+				if ( current_theme_supports( 'experimental-link-color' ) ) {
+					_doing_it_wrong(
+						current_theme_supports( 'experimental-link-color' ),
+						__( '`experimental-link-color` is no longer supported. Use `link-color` instead.', 'gutenberg' ),
+						'6.3.0'
+					);
+				}
 
-			// BEGIN EXPERIMENTAL.
-			// Allow themes to enable appearance tools via theme_support.
-			// This feature was backported for WordPress 6.2 as of https://core.trac.wordpress.org/ticket/56487
-			// and then reverted as of https://core.trac.wordpress.org/ticket/57649
-			// Not to backport until the issues are resolved.
-			if ( current_theme_supports( 'appearance-tools' ) ) {
-				$theme_support_data['settings']['appearanceTools'] = true;
+				// BEGIN EXPERIMENTAL.
+				// Allow themes to enable appearance tools via theme_support.
+				// This feature was backported for WordPress 6.2 as of https://core.trac.wordpress.org/ticket/56487
+				// and then reverted as of https://core.trac.wordpress.org/ticket/57649
+				// Not to backport until the issues are resolved.
+				if ( current_theme_supports( 'appearance-tools' ) ) {
+					$theme_support_data['settings']['appearanceTools'] = true;
+				}
+				// END EXPERIMENTAL.
 			}
-			// END EXPERIMENTAL.
 		}
+
 		$with_theme_supports = new WP_Theme_JSON_Gutenberg( $theme_support_data );
+
+		if ( $options['with_block_style_variations'] ) {
+			$block_style_variations_data = WP_Theme_JSON_Gutenberg::get_from_block_styles_registry();
+			$with_block_style_variations = new WP_Theme_JSON_Gutenberg( $block_style_variations_data );
+			$with_theme_supports->merge( $with_block_style_variations );
+		}
```

Comment:
This change calls the `WP_Theme_JSON_Gutenberg` class to collect and generate theme json data from block style variation data in the block styles registry. That is then merged prior to the theme's theme.json data so that theme.json takes precedence.